### DISCUSSION TITLE
standard_init_linux.go:228: exec user process caused: exec format error after building docker image on MacBook M1 Max for linux/amd64 platform

### DISCUSSION THREAD

**Discussion by skoruba** [dotnet/dotnet-docker#5526] [Q&A] [answered]
### Describe the Bug

When I tried to run image on my server, I got following error and application is not running:

```
standard_init_linux.go:228: exec user process caused: exec format error
```

### Steps to Reproduce

I have following ```DockerFile``` and trying to build this image on my MacBook with M1 Max chip:

```
FROM --platform=$BUILDPLATFORM mcr.microsoft.com/dotnet/aspnet:8.0 AS base
WORKDIR /app
EXPOSE 80
EXPOSE 443

FROM --platform=$BUILDPLATFORM mcr.microsoft.com/dotnet/sdk:8.0 AS build
WORKDIR /src
COPY ["src/MyProject/MyProject.csproj", "src/MyProject/"]
RUN dotnet restore "src/MyProject/MyProject.csproj"
COPY . .
WORKDIR "/src/src/MyProject"
RUN dotnet build "MyProject.csproj" -c Release -o /app/build

FROM build AS publish
RUN dotnet publish "MyProject.csproj" -c Release -o /app/publish

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENV ASPNETCORE_FORWARDEDHEADERS_ENABLED=true
ENTRYPOINT ["dotnet", "MyProject.dll"]
```

For building the image I tried this command:

```
docker buildx build --platform linux/amd64 -t myproject:latest --load -f src/MyProject/Dockerfile .

```

I test architecture after build like this:
```
docker inspect --format '{{.Os}}/{{.Architecture}}' myproject:latest
```

I got ```linux/amd64```

When I tried to run this image on my server, I got following error and application is not running:

```
standard_init_linux.go:228: exec user process caused: exec format error
```

What I tried also:
- Clean old local images
- Check what architecture is on server with command ```uname -m``` and I get ```x86_64```.

### Output of `docker version`

```
Client:
 Cloud integration: v1.0.35+desktop.5
 Version:           24.0.6
 API version:       1.43
 Go version:        go1.20.7
 Git commit:        ed223bc
 Built:             Mon Sep  4 12:28:49 2023
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.24.0 (122432)
 Engine:
  Version:          24.0.6
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.7
  Git commit:       1a79695
  Built:            Mon Sep  4 12:31:36 2023
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.22
  GitCommit:        8165feabfdfe38c65b599c4993d227328c231fca
 runc:
  Version:          1.1.8
  GitCommit:        v1.1.8-0-g82f18fe
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### Output of `docker info`

```
Client:
 Version:    24.0.6
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2-desktop.5
    Path:     /Users/admin/.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.22.0-desktop.2
    Path:     /Users/admin/.docker/cli-plugins/docker-compose
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     /Users/admin/.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.20
    Path:     /Users/admin/.docker/cli-plugins/docker-extension
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v0.1.0-beta.8
    Path:     /Users/admin/.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /Users/admin/.docker/cli-plugins/docker-sbom
  scan: Docker Scan (Docker Inc.)
    Version:  v0.26.0
    Path:     /Users/admin/.docker/cli-plugins/docker-scan
  scout: Docker Scout (Docker Inc.)
    Version:  v1.0.7
    Path:     /Users/admin/.docker/cli-plugins/docker-scout

Server:
 Containers: 8
  Running: 6
  Paused: 0
  Stopped: 2
 Images: 14
 Server Version: 24.0.6
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 8165feabfdfe38c65b599c4993d227328c231fca
 runc version: v1.1.8-0-g82f18fe
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
  cgroupns
 Kernel Version: 6.4.16-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 11
 Total Memory: 7.667GiB
 Name: docker-desktop
 ID: 9e5d9f7f-c0e6-461d-ba1e-7297c73f4fda
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by lbussell** [comment]
Hi @skoruba.

If I'm reading correctly, you want to build your image on your arm64 machine and run it on your amd64 server. You used the `--platform linux/amd64` argument on your Docker build to produce an amd64 image.

There are a couple things to unpack here.

First, a cross-platform build is preferable to building under emulation. What I mean is, your build stage should run as `arm64` but produce `amd64` artifacts, rather than running your entire build as emulated `amd64`. We don't recommend you run **any** .NET tools or builds under Docker emulation as QEMU is not supported by .NET.

The `--platform` switch is only going to affect your **target** platform. When you're running a multi-stage build, Docker implicitly uses the `$TARGETPLATFORM` for your final stage and your final stage only. When you pass `--platform linux/amd64` to your build on your Mac, you are setting `$TARGETPLATFORM=linux/amd64` to `$BUILDPLATFORM=linux/arm64`.

In your Dockerfile, you manually specified `--platform=$BUILDPLATFORM` for both your `base` and `final` layers. This is basically saying you want to ignore the `$TARGETPLATFORM` if you are setting it via the Docker build command.

Take a look at our sample Dockerfile which supports cross-platform builds:

https://github.com/dotnet/dotnet-docker/blob/81f0bdf6696413b814beecf63466bb7230253fa1/samples/aspnetapp/Dockerfile#L1-L22

This Dockerfile supports cross-builds in both directions (`amd64 `-> `arm64` and vice versa). `$BUILDPLATFORM` is used for the build stage, and Docker implicitly picks the `$TARGETPLATFORM` for the final stage. The `dotnet` commands are called with the `-a $TARGETPLATFORM` argument so that it knows to build for a different platform than the one it's building on.

**Reply by skoruba** [reply]
Hi @lbussell 
thank you for your reply, I tried add ```$TARGETARCH``` argument into my dockerfile and try run again:
```docker buildx build --platform linux/amd64 -t myproject:latest --load -f src/MyProject/Dockerfile .```
I still got same error, did I miss something?

**Reply by lbussell** [reply, marked as answer]
Without seeing your updated Dockerfile, it's hard to tell what's going wrong. I think you most likely need to remove the `--platform $BUILDPLATFORM` from your base (runtime) layer. Then it will be implicitly set to `$TARGETPLATFORM` by Docker.

Then your Dockerfile would be:

```Dockerfile
# Remove --platform=$BUILDPLATFORM
FROM mcr.microsoft.com/dotnet/aspnet:8.0 AS base 

WORKDIR /app
EXPOSE 80
EXPOSE 443

FROM --platform=$BUILDPLATFORM mcr.microsoft.com/dotnet/sdk:8.0 AS build

# Add ARG TARGETARCH and `-a $TARGETARCH` build arguments.
ARG TARGETARCH

WORKDIR /src
COPY ["src/MyProject/MyProject.csproj", "src/MyProject/"]
RUN dotnet restore -a $TARGETARCH "src/MyProject/MyProject.csproj"
COPY . .
WORKDIR "/src/src/MyProject"
RUN dotnet build -a $TARGETARCH "MyProject.csproj" -c Release -o /app/build

FROM build AS publish
RUN dotnet publish -a $TARGETARCH "MyProject.csproj" -c Release -o /app/publish

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENV ASPNETCORE_FORWARDEDHEADERS_ENABLED=true
ENTRYPOINT ["dotnet", "MyProject.dll"]
```

**Reply by skoruba** [reply]
This is working, thank you! I had to remove --platform=$BUILDPLATFORM from base and add -a $TARGETARCH
